### PR TITLE
Referees 2018/small refactoring

### PR DESCRIPTION
**Pull request by idanstark42**
This is a little refactoring in order to make to code more evenly split across different files.
The changes are, as stated in the commits: 
 - Splitting the localserver.js into multiple modules in folder server_modules
 - Merging the repeated code in the index.html and admin.html files and their subsequent views, and used session parameters to split into users instead.
 - Added basic log middleware

### CODE REVIEW

File: server_modules/log.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+exports.middleware = function(req, res, next) {
+
+	req.log = function(level, message) {
+		console.log(level + ' - ' + Date.now().toString() + ': ' + message);
```
Comment (poelstra) [inline server_modules/log.js]: Date is now logged as a timestamp. Use `new Date().toISOString()` or `Date().toLocaleString()` instead?

File: server_modules/file_system.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
+var fs = require('fs');
+var Q = require('q');
+var mkdirp = require('mkdirp');
+var path = require('path')
+var utils = require('./utils');
+var args = require('./args');
+
+function getDataFilePath(file) {
+    return exports.resolve(path.dirname(process.argv[1]/*TODO change this to user the args module*/), args.data, file);
```
Comment (poelstra) [inline server_modules/file_system.js]: 1. `args.data` doesn't exist (args calls it `datadir`).

2. Why would the args module have to be used here? The reason for process.argv[1] is to obtain a reference to where the main scripts lives, such that it can stores its data relative to that (in a `data` subdir).

3. The root is already prepended to the path (by `path.dirname(process.argv[1])`), so by using `exports.resolve()` (instead of `path.resolve()`) it now gets added twice. If it is to prevent against "../"-style attacks, this won't work either. (We'd probably need to wrap `file` with an extra call to `path.resolve()` to prevent against this stuff.)

File: server_modules/file_system.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+exports.resolve = function(file) {
+	if(file[0] === '/') {
+		file = utils.root + file;
+	}
+	return path.resolve(file);
+};
+
+exports.readFile = function(file) {
+	if(file[0] === '/') {
```
Comment (poelstra) [inline server_modules/file_system.js]: This now seems asymmetric compared to `writeFile`, so one of them is probably wrong now.
I'm not sure exactly where `readFile` and `writeFile` are used exactly, but it seems a few use-cases are being mixed now: serving files from the `/fs` tree, vs serving generic files from the filesystem.

File: package.json [hunk 4]
```diff
@@ -31,6 +31,7 @@
   "dependencies": {
     "basic-auth-connect": "~1.0.0",
     "express": "^4.10.4",
+    "mustache-express": "^1.2.4",
```
Comment (poelstra) [inline package.json]: Unused dependency.

File: server_modules/file_system.js (hunk 2 above)
Comment (idanstark42) [inline server_modules/file_system.js]: 1. You're right, I changed it.
2. I think it would be more SRP this way, and we already require the args module
3. That was a mistake, and I changed it.
Thank you for the review! Very good comments

File: server_modules/sessions.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+var args = require('./args');
+
+var session = require('express-session');
+var MemoryStore = session.MemoryStore;
+
+var config = {
+	store: new MemoryStore,
```
Comment (poelstra) [inline server_modules/sessions.js]: There is https://www.npmjs.com/package/session-file-store, which looks like a drop-in replacement and saves the session to disk. Can even be nice while developing, when using e.g. forever with watch mode (which will then constantly restart the server whenever a change is detected).

**Comment by poelstra** [comment]
Thanks for the quick response Idan!

Filesystem stuff now works again, "Team initializing..." also disappeared. Nice.

I'm still not happy with the 'asymmetry' in `file_system.js`. Someone using `readFile`, then `writeFile`, is going to have a file read from one location, then saved to another (the data dir).

If you can find the time, it would be best if you could split/rename them in e.g. explicit `readFile`, `readDataFile`, `writeFile`, `writeDataFile` etc. Otherwise, please add a TODO comment with the issues to save future users some hair :)

Rest looks good now, so will merge after that.

(The session store could use improvement, but I won't block the merge on it.)

**Comment by idanstark42** [comment]
So I looked at the fileSystem functions more in depth, And I don't see any differences with which files they go to, could you please explain to me what is the different between them?

**Comment by poelstra** [comment]
Consider the difference between `readFile("/foo.txt")` and `writeFile("/foo.txt", ...)`.
The former will prepend the root, the latter will not.

**Comment by rikkertkoppes** [comment]
wrt mixed indentation: note that there is a .editorconfig file in the root. This works great for your favorite editor: http://editorconfig.org/

**Comment by poelstra** [comment]
Thanks for the quick response @idanstark42, but eh... did you test these changes? ;)

A few remarks:
- `writeFile()` now adds `utils.root`, but when called from the `/fs/` webservice, that service already prepended the dataDir, which is an absolute path.
- `views.js` uses `file_system.resolve()`, which used to do the prefixing trick, but now doesn't anymore: this breaks rendering the front-end
- Because `file_system.resolve()` is now the same as node's built-in `path.resolve()` I think it only makes things more confusing, so probably remove it. Or, put the prefixing back in, but then rename it to reflect that (e.g. `addProgramRoot()`). I think adding the root unconditionally (i.e. without caring about initial slash, like `getDataFilePath()` does) would also help to make it easier to understand.

The 'old' (before this PR) behaviour for readFile, writeFile, etc was that you need to pass it an absolute path already. I think this makes sense; we use different 'roots' in the application, so it's better to be explicit about them. E.g. call `readFile(addStaticHtmlRoot(file))`, or `readFile(getDataFilePath(file))`, or by creating simple wrappers around `readFile()` like `readFileRelative()` vs `readFileFromData()` etc.
I hope I'm making sense :)

**Comment by idanstark42** [comment]
About the `writeFile` in the `/fs/` webservice, yes, your right, that's the point here:
When the `writeFile` receives an absolute path (which start with `C://` or some other Hard drive symbol, but not with `/`), then the `absoluteFile` function doesn't add the root folder. The `absoluteFile` function just makes sure the file is an absolute path. If it is, it doesn't change it. That's why it doesn't affect any files that have come from `getDataFilePath` function
You're right about `file_system.resolve`, I will change that.

**Comment by poelstra** [comment]
Aaaaah, now I see where the confusion is coming from: you're running Windows, I'm running Linux.
On Linux, absolute paths start with a `/`, which broke it pretty badly on my system.

But even with your last changes, it is still broken on Linux, because `writeFile()` now applies a root twice, when called from the `/fs/` webservice.

This is another reason for not conditionally applying the root, but to explicitly always do it where it is needed (especially because different places want a different root).

**Comment by idanstark42** [comment]
Ok, now I get it.
I'll just change it to use path.resolve, that should be cross-platform compatible

File: server_modules/file_system.js [hunk 6]
```diff
@@ -25,11 +25,11 @@ function parseFile(data) {
 }
 
 exports.resolve = function(file) {
-    return path.resolve(absoluteFile(file));
+    return path.resolve(absolutePath(file));
```
Comment (poelstra) [inline server_modules/file_system.js]: `absolutePath()` already does `path.resolve()`, so does one can be removed.

File: server_modules/file_system.js [hunk 7]
```diff
@@ -5,8 +5,8 @@ var path = require('path')
 var utils = require('./utils');
 var args = require('./args');
 
-function absoluteFile(file) {
-    return (file[0] === '/' ? utils.root : '') + file;
+function absolutePath(file) {
+    return path.resolve(path.dirname(process.argv[1]), file);
 }
 
 function getDataFilePath(file) {
```
Comment (poelstra) [inline server_modules/file_system.js]: This function could now use `absolutePath()`, i.e. `return absolutePath(path.join(args.datadir, file))`;

File: server_modules/file_system.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            throw new Error({
+                status: 500,
+                message: 'error reading file'
+            });
+        });
+    });
+}
+
+exports.parseDataFile = function(file) {
+    return exports.parseFile(getDataFilePath(file));
```
Comment (poelstra) [inline server_modules/file_system.js]: `exports` doesn't have `parseFile()` (it's a local function).
Additionally, `parseFile()` expects the file contents, not a filename (I agree the name is confusing, change it if you like).

File: server_modules/file_system.js [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+                res.status(500).send('error reading file');
+                return;
+            }
+        });
+    });
+
+    // writing the "file system"
+    app.post(/^\/fs\/(.*)$/, function(req, res) {
+        var file = getDataFilePath(req.params[0]);
+        exports.writeFile(file, req.body, function(err) {
```
Comment (poelstra) [inline server_modules/file_system.js]: This call to `writeFile()` gets an absolute path through `getDataFilePath()` (which I think is a good thing).
However, `writeFile()` then again calls `absolutePath()`, which seems a bit tricky.

File: server_modules/scores.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        },{});
+    }
+}
+
+exports.route = function(app) {
+
+    //get all, grouped by round
+    app.get('/scores/',function(req,res) {
+        Q.all([
+            fileSystem.parseDataFile('scores.json'),
```
Comment (poelstra) [inline server_modules/scores.js]: This does not do what you think it does. `parseDataFile()` is basically 'convert JSON string to JS object'.

Comment (poelstra) [inline server_modules/scores.js]: (Same holds for all other places where parseDataFile is now used.)

File: server_modules/teams.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
+var utils = require('./utils');
+var fileSystem = require('./file_system');
+
+exports.route = function(app) {
+
+    //get the teams info
+    app.get('/teams',function(req,res) {
+        parseDataFile('teams.json').then(function(result) {
```
Comment (poelstra) [inline server_modules/teams.js]: Undefined reference (same in other places).

File: localserver.js [hunk 12]
```diff
@@ -1,262 +1,42 @@
...
-var resolve = require('path').resolve;
-var argv = require('minimist')(process.argv.slice(2));
-var port = argv.p||1390;
-var basicAuth = require('basic-auth-connect');
-var basicAuthCreds = argv.u;
-var datadir = argv.d||'data';
-var slaveMode = argv.s||false;
+var utils = require('./server_modules/utils');
+var fileSystem = require('./server_modules/file_system');
+var path = require('path');
```
Comment (poelstra) [inline localserver.js]: Unused import.

File: server_modules/args.js [hunk 13]
```diff
@@ -0,0 +1,7 @@
+var argv = require('minimist')(process.argv.slice(2));
+
+exports.port = argv.p || 1390;
+exports.basicAuthCreds = argv.u;
+exports.datadir = argv.d || 'data';
+exports.slaveMode = argv.s || false;
+exports.redisPort = argv.r || 13903;
```
Comment (poelstra) [inline server_modules/args.js]: We don't use Redis?

File: server_modules/challenges.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
+var utils = require('./utils');
+var fileSystem = require('./file_system');
+
+exports.route = function(app) {
+
+    //get challenges over xhr, for hosted service
+    app.get('/challenge/:year', function(req, res) {
+        var file = '/challenges/js/{0}.js'.format(req.params.year);
+
+        fileSystem.readFile(file).then(function(data) {
```
Comment (poelstra) [inline server_modules/challenges.js]: `readFile()` will call `absolutePath`, but because the filename is already absolute, it will try to use `/challenges/...` as-is (i.e. no root).

File: server_modules/scores.js (hunk 10 above)
Comment (idanstark42) [inline server_modules/scores.js]: If that all parseDataFile does, why not just use the javascript JSON.parse built-in function?

File: server_modules/challenges.js (hunk 14 above)
Comment (idanstark42) [inline server_modules/challenges.js]: I just need to remove the '/' in the beginning of the file

File: server_modules/args.js (hunk 13 above)
Comment (idanstark42) [inline server_modules/args.js]: It's there from when I tried to use it for the session, but it didn't work very well...
I think that file based sessions are the best way here, but we need to see how we can get them to work. This will be in a different PR, sense it's its own thing (as I see it)

File: server_modules/scores.js (hunk 10 above)
Comment (idanstark42) [inline server_modules/scores.js]: I just read it in more details... It just add a try-catch mechanism around the JSON parsing, using the JSON.parse function, which I guess I get, since it's server side code. I'll change the name to be clearer

**Comment by idanstark42** [comment]
I'm trying to fix everything... I do want to keep the ideas of absolute path and relative path easy to use, without different functions
Don't worry about the iterations, I'm new to node.js and I love hearing your opinion and advice.

So I tried to create a unified idea of how this module should work, and here's what I got:
1. When writing/reading a file, If the path is absolute, then it will be used, otherwise, the root will be added (This will happen using path.resolve)
2. When reading a JSON file using readJsonFile, The file will return as JSON in a promise ( I saw that the parsing function is only used with the reading function).
3. The fs routing is a get/post system in which you can CRUD a file only in the data folder. 
4. The function getDataFilePath is also public so that it can be used for people wanting to explicitly use the data folder (which currently is always 

And I think I used it correctly in all the relevant places. I tested and it worked on my computer (windows 10). Please let me know if you spot any problems and tell me if you have any comments or places for improvements.

File: server_modules/scores.js (hunk 10 above)
Comment (poelstra) [inline server_modules/scores.js]: Actually, the main reason for the function, is that it returns a Promise for the parsed data, such that it is very easy to use in a Promise chain: `fs.readFile(...).then(parseJSON).then(doSomethingWithTheObject).catch(handleIOorJSorSomethingErrors)`

File: server_modules/args.js (hunk 13 above)
Comment (poelstra) [inline server_modules/args.js]: Agreed, let's get this thing merged first :)
